### PR TITLE
Generic record validation

### LINKED ISSUE #29: Generic record validation and error data structure
### Motivation

https://github.com/the-dr-lazy/cascade/blob/612f19ae91b599da4fa5b724d484ef263e256d69/cascade-api/src/Cascade/Api/Data/User.hs#L59-L105

As it's obvious, making validation functions for product types with record syntax is harsh and verbose but it shouldn't be. We can use the Generic machinery of GHC to make it comfortable.

## Summary

To be able to make a type-level mechanism for making validation function for product types we'll need some type information of the product type like the field name and the value type of the field. The best way to have such information at the type level is row types. Unfortunately, Haskell doesn't have row types yet. So, Generics to the rescue!

### Basic Example

As an example, I'll demonstrate a form that consists of `username` and `future` fields.

Validation of `username` field is totally a simple pure function of type `Text -> Validation ValidationErrors Username` but for the sake of implementation costs we will represent it as follow:

``` haskell
newtype Username = Username { un :: Text }

data ValidationError = IsShort | IsLong | Invalid deriving stock Generic
type ValidationErrors = NonEmpty ValidationError

instance Validatable Username where
  type Raw Username = Text
  type Parsed Username = Username
  type Errors Username = ValidationErrors
  
  validate :: Text -> Sem '[] (Validation ValidationErrors Username)
  validate = undefined
```

As mentioned, there is a potential design space with the `validate @Username` type. As it doesn't need any effects, wrapping the whole validation in `Sem` looks redundant. It's not optimal but makes type-level computations simpler by now. We can reconsider having a more precise type in the future.

Validation of `Future` depends on an effect named `TimeL`:

``` haskell
newtype Future = Future { un :: OffsetDatetime }

data ValidationError = IsPast deriving stock Generic

instance Validatable Future where
  type Raw Future = OffsetDatetime
  type Parsed Future = Future
  type Errors Future = ValidationError
  type Effects Future = '[TimeL]
  
  validate :: OffsetDatetime -> Sem '[TimeL] (Validation ValidationError Future)
  validate = undefined
```

And finally the `Form` type:

``` haskell
data Form (v :: Validity) = Form
  { username :: Validate v Username
  , future   :: Validate v Future
  }
  deriving stock Generic
  deriving Validatable via (Generically (Form v))
```

By deriving `Validatable` for `Form v` we'll be able to validate the product type as follow:

``` haskell
x = validate @(Form 'Raw) Form { username = "user_name", future = 13 }
```

Also, the `Error` type of derived `Validtable` should have `ToJSON` and `FromJSON` instances which render to and un-render from the following JSON structure:

``` json
{ "username": [ { "tag": "IsShort", "message": "username is too short" } ] 
, "future": [ { "tag": "IsPast", "message": "future is not past!" } ]
}
```

### Implementation Details

Please

1. Read [Generic Validation of Nested Data](https://gist.github.com/maksbotan/d358b5ff6081c5d5a94de8d32bbe098c).

Thank you.

### Flags

- [x] I would like to submit a PR if the requested feature becomes approved (Help can be provided if you need assistance submitting a PR.).
- [x] Breaking change.

### PR DESCRIPTION
**Pull request by the-dr-lazy**
This PR resolves #29.
These changes have a great potential to becoming a library on Hackage.

### CODE REVIEW

**Comment by the-dr-lazy** [comment]
This PR is ready for review.

File: cascade-prelude/src/Cascade/Data/Text/NonEmpty.hs [hunk 1]
```diff
@@ -32,3 +38,17 @@ pattern NonEmpty a <- Mk a
 
 mk :: Text -> Maybe NonEmpty
 mk input = if Text.null input then Nothing else Just $ Mk input
+
+data ValidationError = IsEmpty
+  deriving stock (Generic, Data, Show)
+  deriving ToJSON via (ApiErrorFormat ValidationError)
```
Comment (the-dr-lazy) [inline cascade-prelude/src/Cascade/Data/Text/NonEmpty.hs]: This data structure should not be aware about API. Actually this type is an internal type which should not be able to reflect on contracts. Let's defer instance derivation to the `cascade-api` package in a newtype wrapping this one.

File: cascade-api/src/Cascade/Api/Data/Task.hs [hunk 2]
```diff
@@ -10,16 +10,13 @@ Portability : POSIX
 !!! INSERT MODULE LONG DESCRIPTION !!!
 -}
 
+{-# LANGUAGE UndecidableInstances #-}
```
Comment (the-dr-lazy) [inline cascade-api/src/Cascade/Api/Data/Task.hs:13]: Just need an extra line for code styling.

File: cascade-api/src/Cascade/Api/Data/Task.hs [hunk 3]
```diff
@@ -59,98 +52,42 @@ data Readable = Readable
...
-
-  validateDeadlineAt :: Validation RawUpdatableValidationErrors (Maybe Deadline)
-  validateDeadlineAt = case deadlineAt of
-    Just date -> Just <$> case Deadline.mk (unFormattedOffsetDatetime date) now of
-      Nothing -> Failure (mempty { deadlineAt = Just (IsPast :| []) } :: RawUpdatableValidationErrors)
-      Just a  -> Success a
-    Nothing -> Success Nothing
+parseRawUpdatableTask :: Members (Effects (Creatable 'Raw) (Creatable 'Parsed)) r
+                      => Updatable 'Raw
+                      -> Sem r (Validation (Errors (Updatable 'Raw) (Updatable 'Parsed)) (Updatable 'Parsed))
```
Comment (the-dr-lazy) [inline cascade-api/src/Cascade/Api/Data/Task.hs]: Can use `RawUpdatableValidationErrors` type synonym.

File: cascade-api/src/Cascade/Api/Data/Text/Title.hs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  { un :: Text }
+  deriving stock Show
+  deriving newtype (Eq, FromJSON, ToJSON)
+
+makeWrapped ''Title
+
+mk :: Text -> Maybe Title
+mk t = case Text.NonEmpty.mk t of
+  Just _  -> Just $ Mk t
+  Nothing -> Nothing
```
Comment (the-dr-lazy) [inline cascade-api/src/Cascade/Api/Data/Text/Title.hs:36]: This is `coerce . Text.NonEmpty.mk`.

File: cascade-api/src/Cascade/Api/Network/Server/Api/Authentication.hs [hunk 5]
```diff
@@ -41,10 +42,10 @@ import           Validation                          ( validation )
 import           Web.Cookie
 import qualified Web.Cookie                         as Cookie
 
-handleLogin :: Members '[Database.UserL , Error ServerError] r => Authentication.RawCredential -> Sem r LoginResponse
-handleLogin = validation (const $ throw err422) go . parseRawCredential
+handleLogin :: Members '[Database.UserL , Error ServerError] r => Authentication.Credential 'Validation.Raw -> Sem r LoginResponse
+handleLogin = validation (const $ throw err422) go . Authentication.parseRawCredential
```
Comment (jedimahdi) [inline cascade-api/src/Cascade/Api/Network/Server/Api/Authentication.hs]: Login doesn't respond with validation errors, is that intentional?

Comment (the-dr-lazy) [reply cascade-api/src/Cascade/Api/Network/Server/Api/Authentication.hs]: It was. Actually, it was ignored by #17.
BTW, I think it's better to report validation errors.

**Review by jedimahdi — APPROVED** [review]

**Pull request merged**